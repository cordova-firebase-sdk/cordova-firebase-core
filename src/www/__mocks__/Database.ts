import { BaseClass } from "../BaseClass";

export class Database extends BaseClass {
  public app: any;
  private _options: any;
  public _isReady: boolean;

  constructor(app: any, options?: any) {
    super();
    this.app = app;
    this._options = options;
    this._one("fireAppReady", () => {
      this._isReady = true;
    });
  }
};
