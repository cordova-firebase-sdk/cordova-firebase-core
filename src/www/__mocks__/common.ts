export let isInitialized = jest.fn(() => {
  console.log("--->here!");
  return false;
});
export let loadJsPromise = jest.fn();
