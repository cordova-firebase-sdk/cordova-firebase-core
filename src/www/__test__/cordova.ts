export const exec = (onSuccess, onError, pluginName, methodName, args) => {
  onSuccess(args);
};
