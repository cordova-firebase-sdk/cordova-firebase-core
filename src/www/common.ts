// import { Promise } from "es6-promise";

declare let Promise: any;

/**
 * Returns true if given `packageName` is available.
 *
 * ```ts
 * isInitialized("firebase.app"); // true if window.firebase.app is available
 * ```
 * @param packageName - name space
 * @returns true if available.
 */
export const isInitialized = (packageName: string): boolean => {

  let parent: any = window;
  const steps: Array<string> = packageName.split(".");
  const results: Array<string> = steps.filter((step: string): boolean => {
    if (step in parent) {
      parent = parent[step];
      return true;
    } else {
      return false;
    }
  });

  return results.length === steps.length;
};

export interface IloadJsPromiseOptions {

  /**
   * expect JS package name after JS initized.
   */
  package: string;

  /**
   * JS url
   */
  url: string;
}

/**
 * Load JS file, and wait it is initialized.
 * @param options - options
 * @returns Promise<void>
 */
export const loadJsPromise = (options: IloadJsPromiseOptions): Promise<void> => {
  return new Promise((resolve, reject) => {
    if (isInitialized(options.package)) {
      resolve();
    } else {
      const scriptTag: HTMLScriptElement = document.createElement("script") as HTMLScriptElement;
      scriptTag.src = options.url;
      scriptTag.onerror = reject;
      scriptTag.onload = () => {
        let timeout: number = 20;
        const timer = setInterval(() => {
          timeout--;
          if (isInitialized(options.package)) {
            clearInterval(timer);
            resolve();
          }
          if (timeout === 0) {
            clearInterval(timer);
            reject(new Error("[Timeout] JS does not initialized in 10 seconds."));
          }
        }, 500);
      };
      document.body.appendChild(scriptTag);
    }
  });
};

/**
 * @hidden
 */
export const nextTick = (fn) => {
  Promise.resolve().then(fn);
};
