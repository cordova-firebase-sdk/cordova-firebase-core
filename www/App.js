







var utils = require('cordova/utils'),
  common = require('./Common'),
  BaseClass = require('./BaseClass'),
  execCmd = require('./FirebaseCoreCommandQueue');

/*******************************************************************************
 * @name App
 ******************************************************************************/
function App(options, name) {
  BaseClass.call(this);

  var self = this;

  Object.defineProperty(self, 'options', {
    value: options
  });
  Object.defineProperty(self, 'name', {
    value: name
  });

  Object.defineProperty(self, 'id', {
    value: self.hashCode + '_fireapp',
    writable: true,
    enumerable: false
  });
  Object.defineProperty(self, '_isReady', {
    value: false,
    writable: true,
    enumerable: false
  });

  Object.defineProperty(self, '_isRemoved', {
    value: false,
    enumerable: false
  });

  execCmd.call({
    '_isReady': true
  }, function() {
    self._isReady = true;
    self._trigger('ready');
  }, function(error) {
    if (error instanceof Error) {
      throw error;
    } else {
      throw new Error(error);
    }
  }, 'CordovaFirebaseCore', 'newInstance', [{
    'id': self.id,
    'options': options,
    'name': name
  }], {
    'sync': true
  });
}

utils.extend(App, BaseClass);


//---------------------------------------------------------------------------------
// App.database
// https://firebase.google.com/docs/reference/js/firebase.app.App#database
//---------------------------------------------------------------------------------
App.prototype.database = function(url) {
  var self = this;
  if (common.isInitialized('plugin.firebase.database._DBs')) {
    var CordovaFirebaseDatabase = require('cordova-firebase-database.cordova-firebase-database');
    var options = utils.clone(this.options);
    if (url) {
      options.databaseURL = url;
    }
    var db = new CordovaFirebaseDatabase(self.id, options);
    if (self._isReady) {
      db._trigger('fireAppReady');
    } else {
      self._on('ready', function() {
        db._trigger.call(db, 'fireAppReady');
      });
    }
    window.plugin.firebase.database._DBs[db.id] = db;
    return db;
  } else {
    throw new Error('cordova-firebase-database plugin is required to use this feature.');
  }
};







module.exports = App;
