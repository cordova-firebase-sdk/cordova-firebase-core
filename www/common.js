"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var es6_promise_1 = require("es6-promise");
/**
 * Returns true if given `packageName` is available.
 *
 * ```ts
 * isInitialized("firebase.app"); // true if window.firebase.app is available
 * ```
 * @param packageName - name space
 * @returns true if available.
 */
exports.isInitialized = function (packageName) {
    var parent = window;
    var steps = packageName.split(/\\./);
    var results = steps.filter(function (step) {
        if (step in parent) {
            parent = parent[step];
            return true;
        }
        else {
            return false;
        }
    });
    return results.length === steps.length;
};
/**
 * Load JS file, and wait it is initialized.
 * @param options - options
 * @returns Promise<void>
 */
exports.loadJsPromise = function (options) {
    return new es6_promise_1.Promise(function (resolve, reject) {
        if (exports.isInitialized(options.package)) {
            resolve();
        }
        else {
            var scriptTag = document.createElement("src");
            scriptTag.src = options.url;
            scriptTag.onerror = reject;
            scriptTag.onload = function () {
                var timeout = 20;
                var timer = setInterval(function () {
                    timeout--;
                    if (exports.isInitialized(options.package)) {
                        clearInterval(timer);
                        resolve();
                    }
                    if (timeout === 0) {
                        reject(new Error("[Timeout] JS does not initialized in 2 seconds."));
                    }
                }, 10);
            };
            document.body.appendChild(scriptTag);
        }
    });
};
/**
 * @hidden
 */
exports.nextTick = function (fn) {
    es6_promise_1.Promise.resolve().then(fn);
};
