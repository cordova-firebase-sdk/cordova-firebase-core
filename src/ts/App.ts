import { Promise } from "es6-promise";
import { execCmd } from "./CommandQueue";
import { isInitialized } from "./common";
import { PluginBase } from "./PluginBase";

export interface IAppInitializeOptions {

  /**
   * Auth / General Use
   */
  apiKey?: string;

  /**
   * Auth with popup/redirect
   */
  authDomain?: string;

  /**
   * Realtime Database
   */
  databaseURL?: string;

  /**
   * Storage
   */
  storageBucket?: string;

  /**
   * Cloud Messaging
   */
  messagingSenderId?: string;
}

export class App extends PluginBase {

  public name: string = "[DEFAULT]";

  private _options: IAppInitializeOptions = {
    apiKey: null,
    authDomain: null,
    databaseURL: null,
    messagingSenderId: null,
    storageBucket: null,
  };

  private _isInitialized: boolean;

  /**
   * @constructor
   * @param [name] - Application name. Default value is "[DEFAULT]"
   * @param [initOptions] - Application initialize option.
   *  This parameter is for {@link: plugin.firebase.initializeApp()} method.
   */
  constructor(name?: string, initOptions?: IAppInitializeOptions) {
    super("fireapp");

    this._isReady = true;
    this._options = {
      apiKey: initOptions.apiKey || null,
      authDomain: initOptions.authDomain || null,
      databaseURL: initOptions.databaseURL || null,
      messagingSenderId: initOptions.messagingSenderId || null,
      storageBucket: initOptions.storageBucket || null,
    };

    execCmd({
      args: [{
        id: this.id,
        name: this.name,
        options: this._options,
      }],
      context: this,
      methodName: "newInstance",
      pluginName: "CordovaFirebaseCore",
    }).then(() => {
      this._isInitialized = true;
      this._trigger("ready");
    });
  }

  /**
   * Generate FirebaseDatabse instance
   * @param [url] - Realtime database url
   * @returns database instance
   */
  public database(url?: string): any {
    if (isInitialized("plugin.firebase.database._DBs")) {
      const CordovaFirebaseDatabase: any = require("cordvoa-firebase-datbase.cordova-firebase-database");
      const options: any = Object.create(this._options);
      if (url) {
        options.databaseURL = url;
      }
      const db = new CordovaFirebaseDatabase(this.id, options);
      if (this._isInitialized) {
        db._trigger("fireAppReady");
      } else {
        this._on("ready", () => {
          db._trigger("fireAppReady");
        });
      }
    } else {
      throw new Error("cordova-firebase-databse plugin is required");
    }
  }
}
