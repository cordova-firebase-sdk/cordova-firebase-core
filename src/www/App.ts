// import { Promise } from "es6-promise";
import { exec } from "cordova";
import { isInitialized } from "./common";
import { IAppInitializeOptions } from "./IAppInitializeOptions";
import { PluginBase } from "./PluginBase";

declare let Promise: any;
declare let window: any;

export class App extends PluginBase {

  public name: string = "[DEFAULT]";

  private _options: IAppInitializeOptions = {
    apiKey: null,
    authDomain: null,
    databaseURL: null,
    messagingSenderId: null,
    storageBucket: null,
  };

  /**
   * @constructor
   * @param [name] - Application name. Default value is "[DEFAULT]"
   * @param [initOptions] - Application initialize option.
   *  This parameter is for {@link: plugin.firebase.initializeApp()} method.
   */
  constructor(name?: string, initOptions?: IAppInitializeOptions) {
    super("fireapp");

    if (typeof name === "string") {
      this.name = name;
    }

    if (initOptions) {
      this._options = {
        apiKey: initOptions.apiKey || null,
        authDomain: initOptions.authDomain || null,
        databaseURL: initOptions.databaseURL || null,
        messagingSenderId: initOptions.messagingSenderId || null,
        storageBucket: initOptions.storageBucket || null,
      };
    }

    // Create one new instance in native side.
    exec(() => {
      this._isReady = true;
      this._trigger("ready");
    },
    (error: any) => {
      throw new Error(error);
    },
    "CordovaFirebaseCore",
    "newInstance",
    [{
      id: this.id,
      name: this.name,
      options: this._options,
    }]);
  }

  /**
   * Generate FirebaseDatabse instance
   * @param [url] - Realtime database url
   * @returns database instance
   */
  public database(url?: string): any {
    const options: any = this.options;
    if (url) {
      options.databaseURL = url;
    }

    // Load `cordova-firebase-database.Database` module if not yet.
    let database: any;
    if (isInitialized("plugin.firebase.database") &&
      typeof window.plugin.firebase.database === "function") {
      database = window.plugin.firebase.database(this, options);
    } else {
      const moduleName: string = "cordova-firebase-database";

      if (!window.cordova.define.moduleMap[moduleName + ".Database"]) {
        // cordova-firebase-database is not installed.
        throw new Error(moduleName + " plugin is required");
      }
      cordova.require(moduleName + ".Database");
      database = window.plugin.firebase.database(this, options);
    }

    // Waits if native side of FirebaseAppPlugin is not ready yet.
    if (this._isReady) {
      database._trigger("fireAppReady");
    } else {
      this._one("ready", () => {
        database._trigger("fireAppReady");
      });
    }
    return database;
  }

  public get options(): IAppInitializeOptions {
    return JSON.parse(JSON.stringify(this._options));
  }
}


class SecretAppManager {

  public get apps(): Array<App> {
    const keys: Array<string> = Object.keys(this._apps);
    const apps: Array<App> = [];
    keys.forEach((key: string) => {
      apps.push(this._apps[key]);
    });
    return apps;
  }

  public _apps: any = {};

  /**
   * @param initOptions - Application initialize options
   * @param [name] - Application name.
   * @returns Application instance
   */
  public initializeApp(initOptions: IAppInitializeOptions, name?: string): App {
    name = name || "[DEFAULT]";
    if (name in this._apps) {
      throw new Error("Name '" + name + "' application has been already existed.");
    }
    const app: App = new App(name, initOptions);
    this._apps[name] = app;
    return app;
  }
}
if (window.cordova && window.cordova.version) {
  const manager: SecretAppManager = new SecretAppManager();

  const firebaseNS: any = {
    apps: manager.apps,

    ANDROID_SDK_VERSION: "5.5.0",

    database: undefined,

    initializeApp: manager.initializeApp.bind(manager),

    IOS_SDK_VERSION: "5.5.0",

    Promise: Promise.class,

    WEBJS_SDK_VERSION: "5.5.0",
  };

  window.cordova.addConstructor(() => {
    window.plugin = window.plugin || {};
    // (window as any).plugin.firebase = (window as any).plugin.firebase || {};
    if (!window.plugin.firebase) {
      Object.defineProperty(window.plugin, "firebase", {
        value: firebaseNS,
      });
      Object.defineProperty(window.plugin.firebase, "app", {
        value: (name?: string): App => {
          name = name || "[DEFAULT]";
          const app: App = manager._apps[name];
          if (app) {
            return app;
          } else {
            throw new Error("Default app has been not initialized.");
          }
        },
      });
    }
  });
}
