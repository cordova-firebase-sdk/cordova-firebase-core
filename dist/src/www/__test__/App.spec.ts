import { App } from "../App";
import * as common from "../__mocks__/common";
import { Database } from "../__mocks__/Database";
import { exec } from "../__mocks__/cordova";

describe("[App]", () => {

  it ("name should be [DEFAULT]", () => {
    const app: App = new App();
    expect(app.name).toEqual("[DEFAULT]");
  });
  it ("name should be 'hello'", () => {
    const app: App = new App("hello");
    expect(app.name).toEqual("hello");
  });

  describe("[App.databaseURL option]", () => {
    it("should throw error if does not start with https://(projectId).firebaseio.com", () => {
      expect(() => {
        new App("hello", {
          databaseURL: "https://myproject.firebase.com",
        });
      }).toThrowErrorMatchingSnapshot();
    });
    it("should throw error if does not point the root reference", () => {
      expect(() => {
        new App("hello", {
          databaseURL: "https://myproject.firebaseio.com/test/",
        });
      }).toThrowErrorMatchingSnapshot();
    });
    it("should throw error if try to obtain default app before initializing", () => {
      expect(() => {
        (window as any).plugin.firebase.app();
      }).toThrowErrorMatchingSnapshot();
    });
    it("should throw error if specified app has not been initialized", () => {
      expect(() => {
        (window as any).plugin.firebase.app("MYAPP");
      }).toThrowErrorMatchingSnapshot();
    });
    it("should create a DEFAULT app", () => {
      const app = (window as any).plugin.firebase.initializeApp({
        apiKey: "(YOUR API KEY)",
        databaseURL: "https://(PROJECT ID).firebaseio.com",
      });
      expect(app.name).toBe("[DEFAULT]");
    });
    it("should return the DEFAULT app", () => {
      const app = (window as any).plugin.firebase.app();
      expect(app.name).toBe("[DEFAULT]");
    });
    it("should throw error is app name is not string", () => {
      expect(() => {
        (window as any).plugin.firebase.initializeApp({
          apiKey: "(YOUR API KEY)",
          databaseURL: "https://(PROJECT ID).firebaseio.com",
        }, 3333);
      }).toThrowErrorMatchingSnapshot();
    });
    it("should throw error if app is already created", () => {
      expect(() => {
        const app1 = (window as any).plugin.firebase.initializeApp({
          apiKey: "(YOUR API KEY)",
          databaseURL: "https://(PROJECT ID).firebaseio.com",
        }, "MYAPP");
        const app2 = (window as any).plugin.firebase.initializeApp({
          apiKey: "(YOUR API KEY)",
          databaseURL: "https://(PROJECT ID).firebaseio.com",
        }, "MYAPP");
      }).toThrowErrorMatchingSnapshot();
    });
    it("should return 2 apps so far", () => {
      expect((window as any).plugin.firebase.apps).toHaveLength(2);
    });
  });
});

describe("[App.database]", () => {

  // beforeEach(() => {
  //   // Clear all instances and calls to constructor and all methods:
  //   common.isInitialized.mockReset();
  // });

  it("should generate database instance", () => {
    common.isInitialized.mockImplementationOnce(() => {
      return true;
    });
    (window as any).plugin.firebase.database = jest.fn((app, options) => {
      return new Database(app, options);
    });

    const app: App = new App("hello");
    const database = app.database();
    expect(database.app).toEqual(app);
  });

  it("should generate database instance with databaseUrl", () => {
    common.isInitialized.mockImplementationOnce(() => {
      return true;
    });
    (window as any).plugin.firebase.database = jest.fn((app, options) => {
      return new Database(app, options);
    });

    const app: App = new App("hello", {
      apiKey: "(YOUR API KEY)",
      databaseURL: "https://dummy.firebaseio.com",
    });
    const database = app.database("https://dummy.firebaseio.com/users/test");
    expect(database.app).toEqual(app);
  });
  it("should throw Error if additional plugin has not been loaded", () => {
    common.isInitialized.mockImplementationOnce(() => {
      return false;
    });
    (window as any).plugin.firebase.database = jest.fn((app, options) => {
      return new Database(app, options);
    });

    const app: App = new App("hello", {
      apiKey: "(YOUR API KEY)",
      databaseURL: "https://dummy.firebaseio.com",
    });
    expect(() => {
      const database = app.database("https://dummy.firebaseio.com/users/test");
    }).toThrowErrorMatchingSnapshot();
  });
  it("should fire 'fireAppReady' if app fires 'ready' event", (done) => {
    common.isInitialized.mockImplementationOnce(() => {
      return true;
    });
    (window as any).plugin.firebase.database = jest.fn((app, options) => {
      return new Database(app, options);
    });

    const app: App = new App("hello", {
      apiKey: "(YOUR API KEY)",
      databaseURL: "https://dummy.firebaseio.com",
    });
    const database = app.database("https://dummy.firebaseio.com/users/test");
    database._one("fireAppReady", () => {
      setTimeout(() => {
        expect(database.isReady).toBe(true);
        done();
      }, 3);
    });
    app._trigger("ready");
  });
  it("should fire 'fireAppReady' if app has been already initialized", (done) => {
    common.isInitialized.mockImplementationOnce(() => {
      return true;
    });
    (window as any).plugin.firebase.database = jest.fn((app, options) => {
      return new Database(app, options);
    });
    exec.mockReset();
    exec.mockImplementationOnce((onSuccess) => {
      onSuccess();
    });

    const app: App = new App("hello", {
      apiKey: "(YOUR API KEY)",
      databaseURL: "https://dummy.firebaseio.com",
    });

    setTimeout(() => {
      expect(app.isReady).toBe(true);
      const database = app.database("https://dummy.firebaseio.com/users/test");
      setTimeout(() => {
        expect(database.isReady).toBe(true);
        done();
      }, 5);
    }, 10);
  });
});
