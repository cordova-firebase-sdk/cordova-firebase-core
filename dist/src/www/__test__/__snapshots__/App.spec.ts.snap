// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[App.database] should throw Error if additional plugin has not been loaded 1`] = `"cordova-firebase-database plugin is required"`;

exports[`[App] [App.databaseURL option] should throw error if app is already created 1`] = `"Name 'MYAPP' application has been already existed."`;

exports[`[App] [App.databaseURL option] should throw error if does not point the root reference 1`] = `"Database URL must point to the root of a Firebase Database (not including a child path). "`;

exports[`[App] [App.databaseURL option] should throw error if does not start with https://(projectId).firebaseio.com 1`] = `"Cannot parse Firebase url. Please use https://<YOUR FIREBASE>.firebaseio.com"`;

exports[`[App] [App.databaseURL option] should throw error if specified app has not been initialized 1`] = `"App 'MYAPP' has been not initialized."`;

exports[`[App] [App.databaseURL option] should throw error if try to obtain default app before initializing 1`] = `"Default app has been not initialized."`;

exports[`[App] [App.databaseURL option] should throw error is app name is not string 1`] = `"Name must be string."`;
